### PR TITLE
feat: support generation of ES modules based templates

### LINKED ISSUE #539: Support ESM on generate (optional)
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Is there a way to make fastify-cli generate code with esm modules (optional)?

### Motivation

_No response_

### Example

`fastify generate folderName --esm`

### PR DESCRIPTION
**Pull request by nooreldeensalah**
In this PR, I've added support to generation of ES modules based JavaScript template by supplying `--esm` flag, an example would be
```bash
fastify generate folder_name --esm
```

I've written unit tests as well, I've mainly copied `generate.test.js` and did the necessary modifications in a new file `generate-esm.test.js`, to make it easier to review the differences, please refer to this online [diff](https://www.diffchecker.com/aXXnOuz9) between the two files.

~~Finally, I've an issue that I'm not sure how to account for, `npm run test` script runs the tests in the `templates/` directory, but because I've added a new ESM based template, the tests corresponding to these files fail, giving me an error `SyntaxError: Cannot use import statement outside a module`, and I'm not sure what would be the solution to this, because the templates have no `package.json` due to it being generated, therefore, I've made the commits with `--no-verify` flag to bypass the pre-commit hook to be able to discuss the problem with you 😅~~

This PR closes #539
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: templates/app-esm/test/routes/example.test.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+//   t.plan(2)
+//   const app = await build(t)
+//
+//   app.inject({
+//     url: '/example'
+//   }, (err, res) => {
+//     t.error(err)
+//     t.equal(res.payload, 'this is an example')
+//   })
+// })
```
[the remark this answers is not in the document]
Comment (nooreldeensalah) [reply templates/app-esm/test/routes/example.test.js]: Dropped in the next commit

**Review by mcollina — APPROVED** [review]
lgtm

File: templates/app-esm/test/routes/example.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (nooreldeensalah) [reply templates/app-esm/test/routes/example.test.js]: Btw, should I drop all the commented-out blocks, or specifically this file only?

[the remark this answers is not in the document]
Comment (mcollina) [reply templates/app-esm/test/routes/example.test.js]: That's ok

**Comment by mcollina** [comment]
CI is failing

**Comment by nooreldeensalah** [comment]
> CI is failing

CI is indeed failing, and I've mentioned the reason in the last paragraph in my first comment 😓

**Comment by nooreldeensalah** [comment]
The possible solutions I have in mind are:
- Adding a `package.json` file with `type:module`
- Not testing for the tetsts in the `templates/*` directory

Not sure if `node-tap` has some way of allowing to test for ESM files, without a `package.json` file, I've gone through the documentation swiftly, and I couldn't find a way to do this.

File: templates/app-esm/test/routes/example.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (nooreldeensalah) [reply templates/app-esm/test/routes/example.test.js]: I was a bit late to read your comment, but I've already amended the commit to drop the `inject callback style` comments, but I've kept the comment in `support.test.js` (the opts in fastify v2 comment). If there's an issue with that I'll drop it again

**Comment by mcollina** [comment]
You need to add a `package.json` inside `app-esm` with `"type": "module"` in there.

**Comment by nooreldeensalah** [comment]
What do you think of making the argument more clear (i.e., `--type=esm` instead of `--esm`)?
If you do like this suggestion, then I'll modify it accordingly.

File: templates/app-esm/package.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "type": "module"
```
Comment (Eomm) [inline templates/app-esm/package.json:2]: All good, but the generated project prints out:

```
Suites:   ​3 passed​, ​3 of 3 completed​
Asserts:  ​​​3 passed​, ​of 3​
​Time:​   ​920.203ms​
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |       0 |        0 |       0 |       0 |                   
----------|---------|----------|---------|---------|-------------------
```

I recall that there is a configuration to add to tap to let it work

Could you check it?

Comment (nooreldeensalah) [reply templates/app-esm/package.json:2]: Thank you for pointing this out, I've tested the other two templates `app` and `app-ts` and they indeed print the summary properly, not sure what is the issue here, but I'll attempt to check `node-tap` docs.

Comment (nooreldeensalah) [reply templates/app-esm/package.json:2]: After digging through `node-tap` issues and experimenting for a while, it seems like `nyc` doesn't work well with es modules, relevant issues: https://github.com/tapjs/node-tap/issues/735 https://github.com/tapjs/node-tap/issues/813.

Anyway, there seems to be https://github.com/istanbuljs/esm-loader-hook which can be used configured to show coverage for ESM based projects. I've modified the PR accordingly to use this loader.

I'm not sure if this is the optimal solution, but it seems to get the job done. Hopefully `node-tap` will improve this scenario in an upcoming release 🤕

Comment (Eomm) [reply templates/app-esm/package.json:2]: Great work!

```
Suites:   3 passed, 3 of 3 completed
Asserts:  3 passed, of 3
Time:   952.563ms
---------------------|---------|----------|---------|---------|-------------------
File                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
---------------------|---------|----------|---------|---------|-------------------
All files            |     100 |      100 |     100 |     100 |
 ciao                |     100 |      100 |     100 |     100 |
  app.js             |     100 |      100 |     100 |     100 |
 ciao/plugins        |     100 |      100 |     100 |     100 |
  sensible.js        |     100 |      100 |     100 |     100 |
  support.js         |     100 |      100 |     100 |     100 |
 ciao/routes         |     100 |      100 |     100 |     100 |
  root.js            |     100 |      100 |     100 |     100 |
 ciao/routes/example |     100 |      100 |     100 |     100 |
  index.js           |     100 |      100 |     100 |     100 |
---------------------|---------|----------|---------|---------|-------------------
```

**Review by Eomm — APPROVED** [review]

**Pull request merged**